lean/bitvec.lean

```diff
@@ -0,0 +1,375 @@
+                                   apply pow_le_pow_of_le_right, dec_trivial_tac,
+                                   apply le_add_right
+                                 end⟩
+
+  -- bitvec specific version of vector.append
+  def append {m n} (x: bitvec m) (y: bitvec n) : bitvec (m + n)
+    := ⟨ x.val * 2^n + y.val, mul_pow_add_lt_pow x.property y.property ⟩
+
+  def trunc {n : ℕ} (x : bitvec n) (m : ℕ) : bitvec (min n m) :=
+    ⟨ x.val % 2 ^ (min n m), by simp ⟩
+
+  -- splits a bitvector into {n .. m} {m - 1 .. 0} sub bitvectors
+  def split_at {n : ℕ} (m : ℕ) (x : bitvec n) : bitvec (n - m) × bitvec (min n m) := -- upper × lower
+    ⟨ bitvec.cong min_sub_self_right (trunc (ushr x m) (n - m)), trunc x m ⟩
+
+  def split_at' {n m : ℕ} (H : m ≤ n) (x : bitvec n) : bitvec (n - m) × bitvec m := -- upper × lower
+      ⟨ (split_at m x).fst, bitvec.cong (min_eq_right H) ((split_at m x).snd) ⟩
+
+  lemma split_at_snd {n : ℕ} (m : ℕ) (x : bitvec n): (split_at m x).snd.val < 2 ^ m :=
+    begin
+      simp [split_at, trunc],
+      apply nat.lt_of_lt_of_le,
+      { apply mod_lt, apply nat.pos_pow_of_pos, dec_trivial_tac },
+      { apply pow_le_pow_of_le_right,
+        { dec_trivial_tac },
+        { apply min_le_right }
+      }
+    end
+
+  lemma bitvec_intro {n : ℕ} (x y : bitvec n) : x.val = y.val → x = y :=
+  begin
+    intro H,
+    cases x, cases y, congr,
+    simp at H, assumption
+  end
+
+  lemma cong_val {n m : ℕ} {H : n = m} (x : bitvec n)
+  : (bitvec.cong H x).val = x.val :=
+  begin
+    cases x, simp [bitvec.cong]
+  end
+
+  lemma trunc_add_ushr {n m : ℕ} (H : m ≤ n) (x : bitvec n)
+  : (trunc x m).val + (trunc (ushr x m) (n - m)).val * 2 ^ m = x.val :=
+  begin
+    simp [ trunc, min_eq_right H, min_eq_right (sub_le n m), ushr ],
+    rw (pow_mod_superfluous x.property H),
+    { rw mul_comm, apply mod_add_div }
+  end
+
+  lemma split_at'_append_ident {n : ℕ} (m : ℕ) (H : m ≤ n) (x : bitvec n)
+  : append (split_at' H x).fst (split_at'  H x).snd = bitvec.cong (eq.symm (nat.sub_add_cancel H)) x :=
+  begin
+    apply bitvec_intro,
+    simp [append, split_at', cong_val, split_at],
+    apply (trunc_add_ushr H)
+  end
+
+end listlike
+
+section arith
+  -- Arithmetic operations on bitvectors
+  variable {n : ℕ}
+
+  -- 2s complement negation
+  protected def neg (x : bitvec n) : bitvec n :=
+    ⟨if x.val = 0 then 0 else 2^n - x.val,
+     begin
+       by_cases (x.val = 0),
+       { simp [h], apply pos_pow_of_pos, dec_trivial_tac },
+       { simp [h],
+         have pos : 0 < 2^n - x.val, { apply nat.sub_pos_of_lt x.property },
+         have x_val_pos: 0 < x.val, { apply nat.pos_of_ne_zero h },
+         apply sub_lt_of_pos_le x.val (2^n) x_val_pos,
+         apply le_of_lt x.property,
+       }
+      end⟩
+
+  -- Add with carry (no overflow)
+  def adc (x y : bitvec n) (c : bool) : bitvec n × bool :=
+    let c₁ := if c then 1 else 0,
+        r  := x.val + y.val + c₁ in
+    ⟨ ⟨r % 2^n, by simp ⟩, r ≥ 2^n ⟩
```

Comment:
Instead of writing out < r & r&n, by simp >, what if we just had a generic function `of_nat : \Nat -> bitvec n`, that did the rounding and proof.